src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,180 @@
+        tail = newNode;
+        if (last == null) {
+            head = newNode;
+        } else {
+            last.next = newNode;
+        }
+        size++;
+    }
+
+    private void linkBefore(T t, Node<T> beOnTheRight) {
+        final Node<T> prevActual = beOnTheRight.prev;
+        final Node<T> actual = new Node<>(prevActual, t, beOnTheRight);
+        beOnTheRight.prev = actual;
+        if (prevActual == null) {
+            head = actual;
+        } else {
+            prevActual.next = actual;
+        }
+        size++;
+    }
+
+    private Node<T> searchByIndex(int index) {
+        Node<T> search;
+        if (index < (size >> 1)) {
+            search = head;
+            for (int i = 0; i < index; i++) {
+                search = search.next;
+            }
+        } else {
+            search = tail;
+            for (int i = size - 1; i > index; i--) {
+                search = search.prev;
+            }
+        }
+        return search;
+    }
+
     @Override
     public void add(T value) {
+        if (head == null) {
+            linkFirst(value);
+        } else {
+            linkLast(value);
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("Index Out Of Bound");
+        }
+        if (index == size) {
+            linkLast(value);
+        } else {
+            linkBefore(value, searchByIndex(index));
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return searchByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> searchedNode = searchByIndex(index);
+        T oldVal = searchedNode.value;
+        searchedNode.value = value;
+        return oldVal;
+    }
+
+    T unlink(Node<T> unlinkElement) {
+        final T unlinkValue = unlinkElement.value;
+        final Node<T> nextFromUnlink = unlinkElement.next;
+        final Node<T> prevFromUnlink = unlinkElement.prev;
+
+        if (prevFromUnlink == null) {
+            head = nextFromUnlink;
+        } else {
+            prevFromUnlink.next = nextFromUnlink;
+            unlinkElement.prev = null;
+        }
+
+        if (nextFromUnlink == null) {
+            tail = prevFromUnlink;
+        } else {
+            nextFromUnlink.prev = prevFromUnlink;
+            unlinkElement.next = null;
+        }
+
+        unlinkElement.value = null;
+        size--;
+        return unlinkValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        return unlink(searchByIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
```

Comment:
no need to make additional `if-else` block, just put `object == x.value` in your condition